libraries/standard/mqtt/integration-test/mqtt_system_stest.c

```diff
@@ -0,0 +1,718 @@
+
+/**
+ * @brief A valid starting packet ID per MQTT spec. Start from 1.
+ */
+#define MQTT_FIRST_VALID_PACKET_ID          ( 1 )
+
+/**
+ * @brief A PINGREQ packet is always 2 bytes in size, defined by MQTT 3.1.1 spec.
+ */
+#define MQTT_PACKET_PINGREQ_SIZE            ( 2U )
+
+/**
+ * @brief A packet type not handled by MQTT_ProcessLoop.
+ */
+#define MQTT_PACKET_TYPE_INVALID            ( 0U )
+
+/**
+ * @brief Number of milliseconds in a second.
+ */
+#define MQTT_ONE_SECOND_TO_MS               ( 1000U )
+
+/**
+ * @brief Length of the MQTT network buffer.
+ */
+#define MQTT_TEST_BUFFER_LENGTH             ( 128 )
+
+/**
+ * @brief Sample length of remaining serialized data.
+ */
+#define MQTT_SAMPLE_REMAINING_LENGTH        ( 64 )
+
+/**
+ * @brief Subtract this value from max value of global entry time
+ * for the timer overflow test.
+ */
+#define MQTT_OVERFLOW_OFFSET                ( 3 )
+
+/**
+ * @brief Sample topic filter to subscribe to.
+ */
+#define TEST_MQTT_TOPIC                     "/iot/integration/test"
+
+/**
+ * @brief Length of sample topic filter.
+ */
+#define TEST_MQTT_TOPIC_LENGTH              ( sizeof( TEST_MQTT_TOPIC ) - 1 )
+
+/**
+ * @brief Size of the network buffer for MQTT packets.
+ */
+#define NETWORK_BUFFER_SIZE                 ( 1024U )
+
+/**
+ * @brief Client identifier for MQTT session in the tests.
+ */
+#define TEST_CLIENT_IDENTIFIER              "MQTT-Test"
+
+/**
+ * @brief Length of the client identifier.
+ */
+#define TEST_CLIENT_IDENTIFIER_LENGTH       ( sizeof( TEST_CLIENT_IDENTIFIER ) - 1u )
+
+/**
+ * @brief Timeout for receiving CONNACK packet in milli seconds.
+ */
+#define CONNACK_RECV_TIMEOUT_MS             ( 1000U )
+
+/**
+ * @brief Time interval in seconds at which an MQTT PINGREQ need to be sent to
+ * broker.
+ */
+#define MQTT_KEEP_ALIVE_INTERVAL_SECONDS    ( 30U )
+
+/**
+ * @brief Timeout for MQTT_ProcessLoop() function in milliseconds.
+ * The timeout value is appropriately chosen for receiving an incoming
+ * PUBLISH message and ack responses for QoS 1 and QoS 2 communications
+ * with the broker.
+ */
+#define MQTT_PROCESS_LOOP_TIMEOUT_MS        ( 700U )
+
+/**
+ * @brief The MQTT message published in this example.
+ */
+#define MQTT_EXAMPLE_MESSAGE                "Hello World!"
+
+/**
+ * @brief Packet Identifier generated when Subscribe request was sent to the broker;
+ * it is used to match received Subscribe ACK to the transmitted subscribe.
+ */
+static uint16_t globalSubscribePacketIdentifier = 0U;
+
+/**
+ * @brief Packet Identifier generated when Unsubscribe request was sent to the broker;
+ * it is used to match received Unsubscribe ACK to the transmitted unsubscribe
+ * request.
+ */
+static uint16_t globalUnsubscribePacketIdentifier = 0U;
+
+/**
+ * @brief Packet Identifier generated when Publish request was sent to the broker;
+ * it is used to match acknowledgement responses to the transmitted publish
+ * request.
+ */
+static uint16_t globalPublishPacketIdentifier = 0U;
```

Comment:
This is fine for right now, but I could see us having test cases where we have multiple in flight publishes, in which case we'd have to go back and change this code.

Reply:
Sure, this could be upgraded to an array, though it will still be tricky to match the incoming or outgoing packet IDs to their original request IDs as a single call to `ProcessLoop` could involve invocations of `eventCallback` for the multiple different PUBLISHes